### PR TITLE
feat: transcript-extractor module scaffolding (registry + linkSuffixExtractor)

### LINKED ISSUE #427: feat: transcript-extractor module scaffolding (registry + linkSuffixExtractor)
Part of #425

## Context

Add the registry, types, and shared `linkSuffixExtractor` factory under `src/trigger/helpers/transcript-extractors/`. No real extractors yet — the registry is empty but fully dispatchable, so #428 can plug them in one at a time without touching this module. This is intentionally narrow so the surface area can be reviewed for correctness in isolation (lookup, error handling, return shape) before any extractor logic exists.

## Complexity Hint

**Subsystems touched:** trigger helpers (new submodule)
**Independent stories:** 1
**Structural changes needed:** No — new code, no contract changes elsewhere
**Scope note:** none detected

## Task

1. Create the `transcript-extractors/` directory with `types.ts`, `index.ts`, and `link-suffix.ts`.
2. Define `Extractor` and `ExtractorContext` types.
3. Implement the registry as `Map<string /* podcastIndexId */, Extractor>` plus `runPodcastExtractor(ctx)`.
4. Implement `linkSuffixExtractor({ id, suffix, replaceTrailingSlash })` factory using `safeFetch` + `stripHtmlTranscript`.
5. Add unit tests covering: registry miss, registry hit, extractor throws → returns `undefined`, factory variants (`/transcript` append, `-transcript` replace-trailing-slash, missing link → undefined).

## Acceptance Criteria

- [ ] VERIFY: `runPodcastExtractor({ episode, podcast })` returns `undefined` when no extractor is registered for `podcast.podcastIndexId`.
- [ ] VERIFY: `runPodcastExtractor` returns `{ transcript, extractorId }` when a registered extractor returns a non-empty string.
- [ ] VERIFY: When a registered extractor throws, `runPodcastExtractor` logs at warn level and returns `undefined` (does not propagate).
- [ ] VERIFY: `linkSuffixExtractor({ suffix: '/transcript' })` against `https://share.transistor.fm/s/abc123` fetches `https://share.transistor.fm/s/abc123/transcript` (mocked).
- [ ] VERIFY: `linkSuffixExtractor({ suffix: '-transcript', replaceTrailingSlash: true })` against `https://lexfridman.com/jensen-huang/` fetches `https://lexfridman.com/jensen-huang-transcript`.
- [ ] VERIFY: `linkSuffixExtractor` returns `undefined` when `ctx.episode.link` is `null` or empty.
- [ ] VERIFY: All HTTP fetches go through `safeFetch` (SSRF-protected) — never raw `fetch`.
- [ ] VERIFY: HTML responses are passed through `stripHtmlTranscript` before return.

### Files to Create
- `src/trigger/helpers/transcript-extractors/types.ts` — `Extractor`, `ExtractorContext` interfaces.
- `src/trigger/helpers/transcript-extractors/index.ts` — registry `Map` + `runPodcastExtractor`.
- `src/trigger/helpers/transcript-extractors/link-suffix.ts` — `linkSuffixExtractor` factory.
- `src/trigger/helpers/transcript-extractors/__tests__/index.test.ts` — registry dispatch.
- `src/trigger/helpers/transcript-extractors/__tests__/link-suffix.test.ts` — factory variants.

### Approach
1. Define types: `ExtractorContext` carries `{ episode: { podcastIndexId, title, link, rssGuid }, podcast: { podcastIndexId, title } }`. `Extractor` is `{ id: string; extract: (ctx) => Promise<string | undefined> }`.
2. In `index.ts`, declare `const registry = new Map<string, Extractor>();` (left empty in this PR — populated in #428). Export a `register(podcastIndexId, extractor)` helper.
3. Implement `runPodcastExtractor(ctx)`: lookup → if found, run inside try/catch → on success normalize and return `{ transcript, extractorId }`, on miss/empty/throw return `undefined`. Use Trigger.dev `logger.warn` for caught errors.
4. Implement `linkSuffixExtractor`: build URL by either appending or trailing-slash-replacing; `safeFetch` with the existing `FETCH_TIMEOUT_MS` and `MAX_TRANSCRIPT_LENGTH` constants from `src/trigger/helpers/transcript.ts` (lift to a shared constants location if needed).
5. Tests with `vi.mock('@/lib/security', ...)` to stub `safeFetch`.

### Patterns to Follow
- The existing `src/trigger/helpers/transcript.ts` shows the style: pure functions, minimal classes, defensive normalization. Match it.
- Test setup in `src/trigger/__tests__/fetch-transcript.test.ts` for `safeFetch` mocking conventions.

### In Scope
- Registry, types, factory, their unit tests

### Out of Scope
- Specific extractors (Bankless, Lex, Limitless) — handled in #428
- Pipeline integration in `fetch-transcript.ts` — handled in #429
- Admin UI changes — handled in #430

## Dependencies

- Blocked by: #426
- Blocks: #428, #429

## Definition of Done

- [ ] Implementation complete
- [ ] Unit tests passing
- [ ] Lint and build pass
- [ ] PR approved and merged
- [ ] Epic checklist updated

### PR DESCRIPTION
**Pull request by rube-de**
## Summary

Adds the per-podcast transcript-extractor submodule under `src/trigger/helpers/transcript-extractors/` — types, an empty-but-dispatchable registry, and the shared `linkSuffixExtractor` factory. No real extractors and no pipeline integration in this PR (those land in #428 and #429).

- `types.ts` — `Extractor`, `ExtractorContext`, `ExtractorResult`
- `index.ts` — `register`, `runPodcastExtractor`, `__resetRegistry` (single catch-point; `logger.warn` and returns `undefined` on extractor throw)
- `link-suffix.ts` — factory that builds the derived URL (append-suffix or replace-trailing-slash) and delegates the fetch + normalize + truncate pipeline to the existing `fetchTranscriptFromUrl` in `transcript.ts`
- 14 unit tests across `index` (7) and `link-suffix` (7); pipeline-internal cases (HTML stripping, truncation, whitespace-only) live in `transcript.test.ts`

Closes #427

## Test plan

- [x] `bun run lint` — clean
- [x] `bun run format:check` — clean
- [x] `bun run build` — TypeScript compiles
- [x] `bun run build-storybook` — clean
- [x] `bun run test` — 3248/3248 passed (no regression on `transcript.test.ts` or `fetch-transcript.test.ts`)
- [x] All 8 issue VERIFY items mapped to specific test cases (registry miss/hit/throw, factory append, replace-trailing-slash, missing-link, `safeFetch` enforced via `fetchTranscriptFromUrl` reuse, HTML stripped via `fetchTranscriptFromUrl` reuse)
- [x] No raw `fetch`; no `TODO`/`FIXME`/`HACK`/`stub` in the new submodule
- [x] Pre-PR multi-tool review (Codex + pr-review-toolkit + simplify): one §5/§6/§7 finding (`linkSuffixExtractor` duplicated `fetchTranscriptFromUrl`) addressed in-PR by collapsing onto the existing helper; one §6 (use `createTriggerSdkMock`) and one §3 (tighten warn-log assertion) also fixed

### Open Questions

- `index.ts:32` non-Error-throw branch (`String(error)` fallback) is uncovered by tests — only Error throws are exercised. Intentional per code-tester (legitimate edge case); revisit only if a real extractor in #428 surfaces a non-Error reject.
- Pre-PR review flagged secondary improvements (DB-schema brand `PodcastIndexEpisodeId` on the context, `register` throwing on duplicate `podcastIndexId`, `id` as a string-literal union, classifying `AbortError` separately in the warn payload, enriching the warn payload with `episodeLink` / `episode.title`) that are downstream-contract or observability decisions better made in #428/#429 when the first concrete extractors and pipeline caller exist. Not blocking; explicitly out of scope per [docs/code-review-checklist.md §7](docs/code-review-checklist.md).

### Context for Next Session

- Registry is intentionally empty in this PR. #428 plugs in Bankless / Lex / Limitless via `register(podcastIndexId, linkSuffixExtractor({...}))` calls — no further changes needed in this submodule for that. Evidence: `src/trigger/helpers/transcript-extractors/index.ts` (no `register(...)` calls); plan §Boundaries 🔴.
- `linkSuffixExtractor` delegates its fetch path to `fetchTranscriptFromUrl` in `src/trigger/helpers/transcript.ts:160`, so timeout/truncation/HTML-strip behavior changes there will flow through to every future linkSuffix-style extractor automatically. The truncation suffix string `"\n\n[Transcript truncated...]"` lives in three call sites in `transcript.ts` (`fetchTranscript`, `fetchTranscriptFromUrl`, the legacy line in `fetchTranscript`); a follow-up cleanup PR could extract `truncateTranscript` + `TRANSCRIPT_TRUNCATED_SUFFIX` once a non-scaffolding need surfaces — Codex flagged it but it touches paths not introduced by this PR.
- `linkSuffixExtractor` is re-exported from `index.ts` so consumers can `import { linkSuffixExtractor, register } from "@/trigger/helpers/transcript-extractors"` as a single surface in #428. Evidence: `index.ts` re-export line.
- Pipeline wiring (`fetch-transcript.ts` Step 3 between PodcastIndex and description-URL) is deferred to #429 — `runPodcastExtractor` is fully callable today but no caller invokes it. Evidence: `git grep runPodcastExtractor src/` returns the definition only.

## Summary by CodeRabbit

* **Chores**
  * Added internal, pluggable transcript-extractor registry and type contracts to support multiple extraction strategies; Changelog entry updated.
* **New Features**
  * Added a configurable "link suffix" extractor option to transform episode links before fetching transcripts.
* **Tests**
  * Added comprehensive tests covering registry behavior, link-suffix behavior, success/failure cases, trimming, and structured logging on errors.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Chalet-Labs/contentgenie/pull/459)

### CODE REVIEW

File: src/trigger/helpers/transcript-extractors/index.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+export async function runPodcastExtractor(
+  ctx: ExtractorContext,
+): Promise<ExtractorResult | undefined> {
+  const extractor = registry.get(ctx.podcast.podcastIndexId);
+  if (!extractor) return undefined;
+
+  try {
+    const transcript = await extractor.extract(ctx);
+    if (!transcript) return undefined;
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/index.ts]: Fixed: trimmed extractor output with `?.trim()` so whitespace-only results are treated as empty and return `undefined`.

File: src/trigger/helpers/transcript-extractors/index.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    return { transcript, extractorId: extractor.id };
+  } catch (error) {
+    logger.warn("[transcript-extractors] extractor threw", {
+      extractorId: extractor.id,
+      podcastIndexId: ctx.podcast.podcastIndexId,
+      error: error instanceof Error ? error.message : String(error),
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/index.ts]: Fixed: now logging the full error object instead of `error.message` so stack traces and context are preserved.

File: src/trigger/helpers/transcript-extractors/__tests__/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    };
+    register("pod1", extractor);
+    const result = await runPodcastExtractor(makeCtx());
+    expect(result).toBeUndefined();
+    expect(vi.mocked(logger.warn)).toHaveBeenCalledWith(
+      "[transcript-extractors] extractor threw",
+      expect.objectContaining({
+        extractorId: "throws-extractor",
+        podcastIndexId: "pod1",
+        error: "boom",
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/__tests__/index.test.ts:83]: Fixed: updated the error assertion to `expect.any(Error)` (matching the full error object now logged) and added a separate non-Error rejection test case to lock the `String(error)` normalization path.

File: src/trigger/helpers/transcript-extractors/index.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+export async function runPodcastExtractor(
+  ctx: ExtractorContext,
+): Promise<ExtractorResult | undefined> {
+  const extractor = registry.get(ctx.podcast.podcastIndexId);
+  if (!extractor) return undefined;
+
+  try {
+    const transcript = await extractor.extract(ctx);
+    if (!transcript) return undefined;
+    return { transcript, extractorId: extractor.id };
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/index.ts]: Fixed: changed to `(await extractor.extract(ctx))?.trim()` so whitespace-only strings become empty after trim and are returned as `undefined`.

[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/index.ts:40]: Fixed: added `.trim()` before the falsy check so whitespace-only strings (e.g. `"   \n"`) are normalized to empty and returned as `undefined`.

File: src/trigger/helpers/transcript-extractors/index.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import { logger } from "@trigger.dev/sdk";
+import type { Extractor, ExtractorContext, ExtractorResult } from "./types";
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/index.ts]: Fixed: replaced relative imports with `@/trigger/helpers/transcript-extractors/types` and `@/trigger/helpers/transcript-extractors/link-suffix` aliases per AGENTS.md convention.

[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/index.ts]: Fixed: replaced relative imports with `@/trigger/helpers/transcript-extractors/types` and `@/trigger/helpers/transcript-extractors/link-suffix` aliases per AGENTS.md convention.

[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/index.ts]: Fixed: replaced relative imports with `@/trigger/helpers/transcript-extractors/types` and `@/trigger/helpers/transcript-extractors/link-suffix` aliases per AGENTS.md convention.

File: src/trigger/helpers/transcript-extractors/link-suffix.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { fetchTranscriptFromUrl } from "@/trigger/helpers/transcript";
+import type { Extractor, ExtractorContext } from "./types";
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/link-suffix.ts]: Fixed: replaced relative import with `@/trigger/helpers/transcript-extractors/types` alias per AGENTS.md convention.

[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/link-suffix.ts]: Fixed: replaced relative import with `@/trigger/helpers/transcript-extractors/types` alias per AGENTS.md convention.

File: src/trigger/helpers/transcript-extractors/link-suffix.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  return {
+    id: opts.id,
+    extract: async (ctx: ExtractorContext) => {
+      if (!ctx.episode.link) return undefined;
+
+      const base =
+        opts.replaceTrailingSlash && ctx.episode.link.endsWith("/")
+          ? ctx.episode.link.slice(0, -1)
+          : ctx.episode.link;
+      return fetchTranscriptFromUrl(base + opts.suffix);
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/link-suffix.ts]: Fixed: replaced string concatenation with a `URL` constructor — pathname is modified on the parsed URL object so query params and hash are preserved intact.

[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/link-suffix.ts]: Fixed: replaced string concatenation with a `URL` constructor — pathname is modified on the parsed URL object so query params and hash are preserved intact.

File: src/trigger/helpers/transcript-extractors/link-suffix.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+export function linkSuffixExtractor(opts: LinkSuffixOptions): Extractor {
+  return {
+    id: opts.id,
+    extract: async (ctx: ExtractorContext) => {
+      if (!ctx.episode.link) return undefined;
+
+      const base =
+        opts.replaceTrailingSlash && ctx.episode.link.endsWith("/")
+          ? ctx.episode.link.slice(0, -1)
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/link-suffix.ts]: Fixed: added a `rawPath.length > 1` guard so the root-scoped slash is never stripped — avoids the `https://example.com-transcript` malformed-host case.

File: src/trigger/helpers/transcript-extractors/index.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+export type { Extractor, ExtractorContext, ExtractorResult } from "./types";
+export { linkSuffixExtractor } from "./link-suffix";
+
+const registry = new Map<string, Extractor>();
+
+export function register(podcastIndexId: string, extractor: Extractor): void {
+  registry.set(podcastIndexId, extractor);
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/trigger/helpers/transcript-extractors/index.ts:23]: Fixed: added a `registry.has(podcastIndexId)` guard that calls `logger.warn` before overwriting, so duplicate registrations are visible in logs without breaking the existing overwrite behaviour.
